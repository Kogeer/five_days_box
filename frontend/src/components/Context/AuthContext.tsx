import React, {createContext, useEffect, useMemo, useState} from "react";
import {authRepo} from "../../services/api";

interface UserData {
    userName: string;
    userId: string;
}

interface ContextProps {
    isSignedIn: boolean;
    changeIsLoggedIn: (isLogin: boolean, userData: UserData | null) => void;
    userData: UserData | null;
}

export const AuthContext = createContext<ContextProps>({
    isSignedIn: false,
    changeIsLoggedIn: () => {
        return null;
    },
    userData: null
});

const AuthStoreProvider: React.FC = ({children}) => {
    const initialState = (): boolean => !!localStorage.getItem('isSignedIn') || false;

    const [isSignedIn, setIsSignedIn] = useState(initialState());
    const [userData, setUserData] = useState<UserData | null>(null);

    useEffect(() => {
        if (isSignedIn) {
            localStorage.setItem('isSignedIn', 'true');
            return;
        }
        if (!isSignedIn) {
            localStorage.removeItem('isSignedIn');
        }
    }, [isSignedIn, userData]);

    useEffect(() => {
        if (!isSignedIn || !userData) {
            authRepo.me()
                .then(res => {
                    setIsSignedIn(true);
                    setUserData({userName: res.data.name, userId: res.data.id});
                })
                .catch(() => {
                    setIsSignedIn(false);
                });
        }
    }, [isSignedIn, userData]);

    const changeIsLoggedIn = (isLogin: boolean, userData: UserData | null): void => {
        setIsSignedIn(isLogin);
        setUserData(userData);
    }

    const store = useMemo<ContextProps>(() => ({
        isSignedIn,
        userData,
        changeIsLoggedIn
    }), [isSignedIn, userData]);

    return (
        <AuthContext.Provider value={store}>{children}</AuthContext.Provider>
    );
}

export default AuthStoreProvider;
