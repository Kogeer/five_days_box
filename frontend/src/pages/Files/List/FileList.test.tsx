import {MemoryRouter} from "react-router-dom";
import {AuthContext} from "../../../components/Context/AuthContext";
import FileListPage from "./index";
import mock from "../../../test/axios-mock";
import {mockFileResponse} from "../../../test/responses/file-response";
import {act, render, screen, waitFor, within} from "@testing-library/react";

const userData = {userName: 'test user', userId: '123'};

const component = (
    <MemoryRouter>
        <AuthContext.Provider value={{userData, isSignedIn: true, changeIsLoggedIn: (): void => {}}}>
            <FileListPage />
        </AuthContext.Provider>
    </MemoryRouter>
);

describe('FileList', () => {
    it('Render files in table', async () => {
        mock.onGet('/files').reply(200,  mockFileResponse);

        const {findByText, asFragment} = render(component);

        const fileText = await findByText('fake file');

        await waitFor(() => {
            expect(fileText).toBeInTheDocument();
        });
        expect(asFragment()).toMatchSnapshot();
    });

    it('Not show actions if user not own the file', async () => {
        mock.onGet('/files').reply(200,  mockFileResponse);

        await act(async () => {
            render(component)
        });

        const fileRow = (screen.getByText('another fake file').closest('tr') as HTMLTableRowElement);
        const utils = within(fileRow);

        await waitFor(() => {
            expect(utils.queryByText('Download')).not.toBeInTheDocument();
            expect(utils.queryByText('Delete')).not.toBeInTheDocument();
        });
    });
});
