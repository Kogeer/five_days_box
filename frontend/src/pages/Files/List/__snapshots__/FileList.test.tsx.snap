// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileList Render files in table 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column align-items-end container"
  >
    <a
      href="/files/upload"
    >
      <button
        class="mb-3 btn btn-primary"
        type="button"
      >
        Upload
      </button>
    </a>
    <table
      class="table table-striped table-bordered table-hover"
    >
      <thead>
        <tr>
          <th>
            File name
          </th>
          <th>
            Upload time
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="align-middle"
        >
          <td>
            fake file
          </td>
          <td>
            10/28/2021, 1:04:20 PM
          </td>
          <td>
            <div
              class="btn-group"
              role="group"
            >
              <button
                class="btn btn-primary"
                type="button"
              >
                Share
              </button>
              <button
                class="btn btn-success"
                type="button"
              >
                Download
              </button>
              <button
                class="btn btn-danger"
                type="button"
              >
                Delete
              </button>
            </div>
          </td>
        </tr>
        <tr
          class="align-middle"
        >
          <td>
            another fake file
          </td>
          <td>
            10/28/2021, 1:04:20 PM
          </td>
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
